### PR TITLE
Xcode 10.2 b1 GameController APIs

### PR DESCRIPTION
**Pull request by chamons**
See [diff](https://gist.github.com/chamons/aa8691ec3d900744675022a6a10a7e3c) for GCExtendedGamepadSnapshotData since it's an extension of existing tyype.

See [diff](https://gist.github.com/chamons/092ae1a0efbc90abea4664a98960e133) for GCMicroGamepadSnapshotData as well.

### CODE REVIEW

File: src/GameController/GCMicroGamepadSnapshot.cs [hunk 1]
```diff
@@ -48,11 +82,25 @@ public partial class GCMicroGamepadSnapshot {
...
+		[Mac (10, 12)]
+		public static bool TryGetSnapshotData (NSData data, out GCMicroGamepadSnapshotData snapshotData)
+		{
+			return GCMicroGamepadSnapshotDataFromNSData (out snapshotData, data == null ? IntPtr.Zero : data.Handle);
+		}
+
 	}
 }
 
-#endif // IOS || TVOS
```
Comment (chamons) [inline src/GameController/GCMicroGamepadSnapshot.cs]: The comment here is just wrong, the block is `#if !WATCHOS` above.

**Comment by chamons** [comment]
I also am unsure on the availability attributes for the new types. I'd think they were 10.14.4 and equivalent, since it's new, but the header doesn't help me here...

File: src/GameController/GCMicroGamepadSnapshot.cs [hunk 2]
```diff
@@ -48,11 +82,25 @@ public partial class GCMicroGamepadSnapshot {
 		static extern bool GCMicroGamepadSnapShotDataV100FromNSData (out GCMicroGamepadSnapShotDataV100 snapshotData, /* NSData */ IntPtr data);
 
 		[Mac (10, 12)]
+		[Deprecated (PlatformName.iOS, message: "Use 'TryGetSnapshotData (NSData, out GCMicroGamepadSnapshotData)' instead.")]
+		[Deprecated (PlatformName.MacOSX, message: "Use 'TryGetSnapshotData (NSData, out GCMicroGamepadSnapshotData)' instead.")]
+		[Deprecated (PlatformName.TvOS, message: "Use 'TryGetSnapshotData (NSData, out GCMicroGamepadSnapshotData)' instead.")]
```
Comment (spouliot) [inline src/GameController/GCMicroGamepadSnapshot.cs]: ^ this does not mention the version where the API are deprecated, which is important if you don't target the latest

File: src/GameController/GCMicroGamepadSnapshot.cs [hunk 3]
```diff
@@ -14,6 +14,9 @@ namespace GameController {
 	// GCMicroGamepadSnapshot.h
 	// float_t are 4 bytes (at least for ARM64)
 	[StructLayout (LayoutKind.Sequential, Pack = 1)]
+	[Deprecated (PlatformName.iOS, message: "Use 'GCMicroGamepadSnapshotData' instead.")]
+	[Deprecated (PlatformName.MacOSX, message: "Use 'GCMicroGamepadSnapshotData' instead.")]
+	[Deprecated (PlatformName.TvOS, message: "Use 'GCMicroGamepadSnapshotData' instead.")]
```
Comment (spouliot) [inline src/GameController/GCMicroGamepadSnapshot.cs]: no version mentioned

File: src/GameController/GCExtendedGamepadSnapshot.cs [hunk 4]
```diff
@@ -71,10 +118,24 @@ public partial class GCExtendedGamepadSnapshot {
...
+		
+		[DllImport (Constants.GameControllerLibrary)]
+		static extern bool GCExtendedGamepadSnapshotDataFromNSData (
+			/* GCExtendedGamepadSnapshotData * __nullable */ out GCExtendedGamepadSnapshotData snapshotData, 
+			/* NSData * __nullable */ IntPtr data);
+
 
+		[Deprecated (PlatformName.iOS, message: "Use 'TryGetExtendedSnapShotData' instead.")]
+		[Deprecated (PlatformName.MacOSX, message: "Use 'TryGetExtendedSnapShotData' instead.")]
+		[Deprecated (PlatformName.TvOS, message: "Use 'TryGetExtendedSnapShotData' instead.")]
```
Comment (spouliot) [inline src/GameController/GCExtendedGamepadSnapshot.cs]: no version mentioned

File: src/GameController/GCExtendedGamepadSnapshot.cs [hunk 5]
```diff
@@ -63,6 +60,56 @@ public NSData ToNSData ()
 			return p == IntPtr.Zero ? null : new NSData (p);
 		}
 	}
+	
+	// float_t are 4 bytes (at least for ARM64)
+	[StructLayout (LayoutKind.Sequential, Pack = 1)]
+	[iOS (7,0)]
+	[Mac (10,9, onlyOn64: true)]
```
Comment (spouliot) [inline src/GameController/GCExtendedGamepadSnapshot.cs]: fine for the stuct

File: src/GameController/GCExtendedGamepadSnapshot.cs [hunk 6]
```diff
@@ -63,6 +60,56 @@ public NSData ToNSData ()
...
+
+		[TV (12, 2), Mac (10, 14, 4, onlyOn64: true), iOS (12, 2)]
+		bool LeftThumbstickButton;
+		
+		[TV (12, 2), Mac (10, 14, 4, onlyOn64: true), iOS (12, 2)]
+		bool RightThumbstickButton;
+
+		[DllImport (Constants.GameControllerLibrary)]
+		static extern /* NSData * __nullable */ IntPtr NSDataFromGCExtendedGamepadSnapshotData (
+			/* GCExtendedGamepadSnapshotData * __nullable */ ref GCExtendedGamepadSnapshotData snapshotData);
```
Comment (spouliot) [inline src/GameController/GCExtendedGamepadSnapshot.cs]: that API did not exists before Xcode 10.2 right ?
so it's missing it's availability

File: src/GameController/GCExtendedGamepadSnapshot.cs [hunk 7]
```diff
@@ -63,6 +60,56 @@ public NSData ToNSData ()
...
+
+		public NSData ToNSData ()
```
Comment (spouliot) [inline src/GameController/GCExtendedGamepadSnapshot.cs]: same (availability inherited from struct)

File: src/GameController/GCExtendedGamepadSnapshot.cs [hunk 8]
```diff
@@ -71,10 +118,24 @@ public partial class GCExtendedGamepadSnapshot {
 		static extern bool GCExtendedGamepadSnapShotDataV100FromNSData (
 			/* GCExtendedGamepadSnapShotDataV100 * __nullable */ out GCExtendedGamepadSnapShotDataV100 snapshotData, 
 			/* NSData * __nullable */ IntPtr data);
+		
+		[DllImport (Constants.GameControllerLibrary)]
+		static extern bool GCExtendedGamepadSnapshotDataFromNSData (
+			/* GCExtendedGamepadSnapshotData * __nullable */ out GCExtendedGamepadSnapshotData snapshotData, 
+			/* NSData * __nullable */ IntPtr data);
```
Comment (spouliot) [inline src/GameController/GCExtendedGamepadSnapshot.cs]: ^ no availability (intro will complain on older bots)

File: src/GameController/GCExtendedGamepadSnapshot.cs [hunk 9]
```diff
@@ -71,10 +118,24 @@ public partial class GCExtendedGamepadSnapshot {
...
 
+		[Deprecated (PlatformName.iOS, message: "Use 'TryGetExtendedSnapShotData' instead.")]
+		[Deprecated (PlatformName.MacOSX, message: "Use 'TryGetExtendedSnapShotData' instead.")]
+		[Deprecated (PlatformName.TvOS, message: "Use 'TryGetExtendedSnapShotData' instead.")]
 		public static bool TryGetSnapShotData (NSData data, out GCExtendedGamepadSnapShotDataV100 snapshotData)
 		{
 			return GCExtendedGamepadSnapShotDataV100FromNSData (out snapshotData, data == null ? IntPtr.Zero : data.Handle);
 		}
+		
+		public static bool TryGetExtendedSnapShotData (NSData data, out GCExtendedGamepadSnapshotData snapshotData)
```
Comment (spouliot) [inline src/GameController/GCExtendedGamepadSnapshot.cs]: should have the same availability as the p/invoke it calls

File: src/GameController/GCMicroGamepadSnapshot.cs [hunk 10]
```diff
@@ -40,6 +43,37 @@ public NSData ToNSData ()
...
+		public float /* float_t = float */ DPadX;
+		public float /* float_t = float */ DPadY;
+
+		// Buttons in the range [0.0, 1.0]
+		public float /* float_t = float */ ButtonA;
+		public float /* float_t = float */ ButtonX;
+
+		[DllImport (Constants.GameControllerLibrary)]
+		static extern /* NSData * __nullable */ IntPtr NSDataFromGCMicroGamepadSnapshotData (
+			/* __nullable */ ref GCMicroGamepadSnapshotData snapshotData);
```
Comment (spouliot) [inline src/GameController/GCMicroGamepadSnapshot.cs]: if new in Xcode 10.2 then it needs availability attributes

File: src/GameController/GCMicroGamepadSnapshot.cs [hunk 11]
```diff
@@ -40,6 +43,37 @@ public NSData ToNSData ()
...
+
+		public NSData ToNSData ()
```
Comment (spouliot) [inline src/GameController/GCMicroGamepadSnapshot.cs]: same ones here

File: src/GameController/Enums.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+	public enum GCExtendedGamepadSnapshotDataVersion : long
+	{
+		Version1 = 0x0100,
+		Version2 = 0x0101,
+	}
+	
+	[Native]
+	public enum GCMicroGamepadSnapshotDataVersion : long
+	{
+		Version1 = 0x0100
```
Comment (rolfbjarne) [inline src/GameController/Enums.cs]:
```suggestion
		Version1 = 0x0100,
```

**Pull request merged**